Add coreClient to DomainClient

aynleslie opened the pull request:
#### What it do?
This pipes the core client through to the domain client.   This will enable looking up user info so that we can optionally hide locked domains (in my next commit likely).  Note the client isn't actually used anywhere yet, just created and passed along. 

This adds TestHttpClient and TestCoreClient to facilitate easier testing, since so many tests create DomainClients.  

I fixed a bug I introduced in the perf code.

#### Notes:

1. I've already pushed the corresponding changes that will be needed in apps-marathon in commit a085fe040dfae7433b to master.   Don't worry about them preceding this branch since they just add more env variables.
2. I've already made and clortho'd the app-tokens.  They reside at /disco/cetera-apptoken.conf in the clortho bucket.


#### How was it tested. 

sbt styleCheck test run (on both projects)

rlvoyer commented on cetera-http/src/test/scala/com/socrata/cetera/authentication/CoreClientSpec.scala:
```diff
@@ -101,15 +92,15 @@ class CoreClientSpec extends WordSpec with ShouldMatchers with BeforeAndAfterAll
     }
 
-    "return the anonymous user without calling core if passed no cookie" in {
+    "return None without calling core if passed no cookie" in {
```
Why do we even want to call fetchCurrentUser if there is no cookie?

jasonij replied:
fetchCurrentUser is not a great name. It implies that we're maintaining state on Cetera about the current_user object. Really, we're just taking a cookie, validating it against core, and getting a user object back. validateAndFetchUser or validateCookieAndFetchUser or  something else would be better. (I mentioned this before but GH dropped my comments evidently.)

aynleslie replied:
voyer: pulling out the cookie header gives an option, so I'm testing all cases. The logic of dealing with the None has to go somewhere - I placed it in the fetchCurrentUser method.

jason:  we called all of the go-get-X-from-core methods as fetchX in rammy.  I was just following that practice, since people hate getX.  I disagree it sounds like we are maintaining state - this method resides in the core client.  I'm happy to rename it, though not now since I'd rather not add a refactor to this commit - I'm trying to keep each commit small and focused.  However, the next commit will touch on that code anyway, so I'll change it then.

jasonij replied:
My point is that this is more than just fetching, it's a validation of a user session token or something like that, and it happens per request (or as needed) rather than being internal state. `current_user` has connotations.

rlvoyer commented on cetera-http/src/test/scala/com/socrata/cetera/authentication/CoreClientSpec.scala:
```diff
@@ -160,14 +151,11 @@ class CoreClientSpec extends WordSpec with ShouldMatchers with BeforeAndAfterAll
 
   "When core is not reachable, the isCookieValid method" should {
-    "return the anonymous user" in {
+    "return None" in {
```
Are you testing the `isCookieValid` method or the `fetchCurrentUser` method? This test is slightly confusing.

aynleslie replied:
doh, guess we missed that.  will fix.

jasonij replied:
Really, the cookie is the thing holding the token. The question is not whether the cookie is valid, but whether the token itself is valid.

rlvoyer commented on cetera-http/src/test/scala/com/socrata/cetera/services/SearchServiceSpec.scala:
```diff
@@ -25,13 +25,16 @@ import com.socrata.cetera.util.Params
-  val domainClient: DomainClient = new DomainClient(client, testSuiteName)
+  val httpClient = new TestHttpClient()
+  val coreClient = new TestCoreClient(httpClient, 8036)
+  val domainClient: DomainClient = new DomainClient(client, coreClient, testSuiteName)
```
The type annotation here is unnecessary.

rlvoyer commented on cetera-http/src/test/scala/com/socrata/cetera/services/FacetServiceSpec.scala:
```diff
@@ -6,11 +6,13 @@ import org.scalatest.{BeforeAndAfterAll, FunSuiteLike, Matchers}
-  val domainClient: DomainClient = new DomainClient(client, testSuiteName)
+  val httpClient = new TestHttpClient()
+  val coreClient = new TestCoreClient(httpClient, 8035)
+  val domainClient: DomainClient = new DomainClient(client, coreClient, testSuiteName)
```
The type annotation here is unnecessary.

rlvoyer commented on cetera-http/src/test/scala/com/socrata/cetera/services/DomainCountServiceSpec.scala:
```diff
@@ -5,11 +5,13 @@ import org.scalatest.{BeforeAndAfterAll, FunSuiteLike, Matchers}
-  val domainClient: DomainClient = new DomainClient(client, testSuiteName)
+  val httpClient = new TestHttpClient()
+  val coreClient = new TestCoreClient(httpClient, 8034)
+  val domainClient: DomainClient = new DomainClient(client, coreClient, testSuiteName)
```
The type annotation here is unnecessary.

rlvoyer commented on cetera-http/src/test/scala/com/socrata/cetera/services/DomainBoostSpec.scala:
```diff
@@ -10,7 +10,9 @@ import com.socrata.cetera.util.Params
-  val domainClient: DomainClient = new DomainClient(client, testSuiteName)
+  val httpClient = new TestHttpClient()
+  val coreClient = new TestCoreClient(httpClient, 8030)  // Remember to close() me!!
+  val domainClient: DomainClient = new DomainClient(client, coreClient, testSuiteName)
```
The type annotation here is unnecessary.

rlvoyer commented on cetera-http/src/test/scala/com/socrata/cetera/services/DatatypeBoostSpec.scala:
```diff
@@ -14,7 +14,9 @@ class DatatypeBoostSpec extends FunSuiteLike with Matchers with TestESData with
-  val domainClient: DomainClient = new DomainClient(client, testSuiteName)
+  val httpClient = new TestHttpClient()
+  val coreClient = new TestCoreClient(httpClient, 8033)
+  val domainClient: DomainClient = new DomainClient(client, coreClient, testSuiteName)
```
The type annotation here is unnecessary.

rlvoyer commented on cetera-http/src/test/scala/com/socrata/cetera/services/CountServiceSpec.scala:
```diff
@@ -112,7 +115,9 @@ class CountServiceSpec extends FunSuiteLike with Matchers with BeforeAndAfterAll
-  val domainClient: DomainClient = new DomainClient(client, testSuiteName)
+  val httpClient = new TestHttpClient()
+  val coreClient = new TestCoreClient(httpClient, 8032)
+  val domainClient: DomainClient = new DomainClient(client, coreClient, testSuiteName)
```
The type annotation here is unnecessary.

aynleslie replied:
really, no idea how this happened.  will fix all of these.

sokrahta replied:
each of the value type annotations was there before this change, thanks for cleaning it up Ayn!

rlvoyer commented on cetera-http/src/test/scala/com/socrata/cetera/services/CountServiceSpec.scala:
```diff
@@ -5,20 +5,23 @@ import com.rojoma.json.v3.codec.DecodeError
+  val esClient: ElasticSearchClient = new TestESClient(testSuiteName)
+  val httpClient = new TestHttpClient()
+  val coreClient = new TestCoreClient(httpClient, 8031)
+  val domainClient: DomainClient = new DomainClient(esClient, coreClient, testSuiteName)
```
Why the type annotations?

aynleslie replied:
ya know.  I have no idea.  I don't recall writing that. will fix.

rlvoyer commented on cetera-http/src/test/scala/com/socrata/cetera/services/CountServiceSpec.scala:
```diff
@@ -5,20 +5,23 @@ import com.rojoma.json.v3.codec.DecodeError
+  val httpClient = new TestHttpClient()
+  val coreClient = new TestCoreClient(httpClient, 8031)
+  val domainClient: DomainClient = new DomainClient(esClient, coreClient, testSuiteName)
+  val documentClient: DocumentClient = new DocumentClient(esClient, domainClient, testSuiteName, None, None, Set.empty)
```
Ditto.

rlvoyer commented:
One question about the CoreClient: why is cookie optional? If we need a cookie to get a user, why not make it required and handle the logic in the caller (ie. outside of the CoreClient) regarding to call fetchCurrentUser?

jasonij commented on cetera-http/src/main/resources/reference.conf:
```diff
@@ -29,6 +29,12 @@ com.socrata {
+  core {
+    host = "127.0.0.1"
+    port = 8081
+    appToken = "REDACTED"
```
So, Cetera won't work if the appToken is missing. Try deleting this line and running.

aynleslie replied:
yes, I didn't write the coreClient to optionally take an appToken. I could have, but we really should always use one.

jasonij commented:
I agree with being much tighter about Options. Any time we are logically safe in getting rid of Options, we should do so. If we can't be rid of an Option, we should at least push the Option back as far as possible, so de-option as soon as possible.

sokrahta commented on cetera-http/src/test/scala/com/socrata/cetera/search/DomainClientSpec.scala:
```diff
@@ -16,6 +18,7 @@ class DomainClientSpec extends WordSpec with ShouldMatchers with TestESData with
     removeBootstrapData()
     client.close()
+    httpClient.close()
   }
```
remember to `coreClient.close()`

aynleslie replied:
the core client doesn't need to be closed.  It's the underlying httpClient that needs to be closed.

sokrahta commented on cetera-perf/src/main/scala/com/socrata/cetera/PerfESClient.scala:
```diff
@@ -83,6 +83,8 @@ class PerfESClient(testSuiteName: String = "catalog")
       Some(Random.alphanumeric.take(16).force.mkString),
       Random.nextBoolean(),
+      Random.nextBoolean(),
+      Random.nextBoolean(),
```
:+1: thanks for paying attention to the perf test even though the normal build process ignores it!

jasonij commented on cetera-http/docker/cetera.conf.j2:
```diff
@@ -29,6 +29,12 @@ com.socrata {
+  core {
+    host = {{ CORE_HOST }}
+    port = {{ CORE_PORT }}
+    {{ CORE_APP_TOKEN_LINE }}
```
If the appToken is required, then we should follow appToken = {{ CORE_APP_TOKEN }} or whatever.

My preference would be, since this makes it look like it is unnecessary, to support not having it.

But in any event we need to be consistent. Will the appToken be required?

aynleslie replied:
The trade off there is that you then have to source in the variable from the /dev/shm directory, which breaks local docker builds.  I took that approach in animalysis and it made the docker files and run scripts kind of gross.  

And honestly, I don't see a problem with the deploy choking if we can't properly reach the clortho bucket to pull it out.   A choked deploy doesn't cause a service interruption and in the case the token is missing, we should add it, not ignore the problem by optionally using an app-token.

aynleslie commented:
re optioning the cookie:  Pulling a header out of a req returns an Option.  We have to deal with it somewhere.  The logic of "are you validated" currently falls on the coreClient, so I put it there.  If y'all would rather see that moved to the Domain Client, I can do so.

aynleslie commented:
I pushed changes to take away the type annotations (there were several others btw) and fixed the typo in the test description.   This leaves three issues y'all have brought up:

1. changing the name of the core method.  My next commit has to add another method to the coreClient, so I will address that then.  
2. dealing with the option higher up.  In this case, the logic would go into the domainClient. I still think that it makes more sense in the core client. The frontend shall send us a cookie  - maybe that's a None or empty or a bad cookie or a good cookie.  Regardless, we are supposed to pass that to core, who makes the decision and sends us back a 200 or 401 or whatever.  But we already know what core's response is for Nones and empty strings, so rather than call core, we deal with it.  It is an optimization to not bother core.  It is the coreClient's job to figure out auth, not the domainClient.  
3. changing the way we bring in the app token to use jason's suggestion. Katie and I attempted to use that strategy when we added the app-token in rammy.  We ran into problems with the env variables not being in scope at the right time.  I've gotten around this with hacky solutions in the dockerfile and run scripts in the past, but slemko suggested that we use the method you see in this commit.  It's much cleaner. Regarding the safety of this choice, we already created app-tokens and they are already in place.  We could add a default to the config class (similar to what was done with the timeouts) in order to avoid failed deploys in the case clortho is having issues, but I feel this hides the problem rather than addressing it.

rlvoyer commented on cetera-http/src/test/scala/com/socrata/cetera/services/SearchServiceSpec.scala:
```diff
@@ -225,12 +228,14 @@ class SearchServiceSpec extends FunSuiteLike with Matchers with BeforeAndAfterAl
+  val coreClient = new TestCoreClient(httpClient, 8037)
+  val domainClient = new DomainClient(client, coreClient, testSuiteName)
+  val documentClient = new DocumentClient(client, domainClient, testSuiteName, None, None, Set.empty)
+  val balboaDir = new java.io.File("balboa_test_trash")
```
No biggie, but why not just import java.io.File?

rlvoyer commented:
:shipit:

aynleslie commented:
manually merged.  will address things soon.
